### PR TITLE
Scroll bar position fix

### PR DESCRIPTION
**Pull request by dodler**
Fixed scroll bar position via adding map for keeping values of scroll bar for different models.

### CODE REVIEW

**Comment by dodler** [comment]
https://bugs.launchpad.net/mixxx/+bug/1450391

Bug on launchpad.

**Comment by Be-ing** [comment]
I tried to test and got a compile error:
```
src/widget/wanalysislibrarytableview.cpp: In constructor ‘WAnalysisLibraryTableView::WAnalysisLibraryTableView(QWidget*, UserSettingsPointer, TrackCollection*)’:
src/widget/wanalysislibrarytableview.cpp:7:86: error: no matching function for call to ‘WTrackTableView::WTrackTableView(QWidget*&, UserSettingsPointer&, TrackCollection*&, QString)’
         : WTrackTableView(parent, pConfig, pTrackCollection, QString::number(qrand())) {
                                                                                      ^
In file included from src/widget/wanalysislibrarytableview.h:7:0,
                 from src/widget/wanalysislibrarytableview.cpp:2:
src/widget/wtracktableview.h:30:5: note: candidate: WTrackTableView::WTrackTableView(QWidget*, UserSettingsPointer, TrackCollection*, bool)
     WTrackTableView(QWidget* parent, UserSettingsPointer pConfig,
     ^~~~~~~~~~~~~~~
src/widget/wtracktableview.h:30:5: note:   no known conversion for argument 4 from ‘QString’ to ‘bool’
scons: *** [lin64_build/widget/wanalysislibrarytableview.o] Error 1
scons: building terminated because of errors.
```

File: src/widget/wlibrarytableview.cpp [hunk 1]
```diff
@@ -42,7 +42,7 @@ WLibraryTableView::WLibraryTableView(QWidget* parent,
     setVerticalScrollBarPolicy(Qt::ScrollBarAlwaysOn);
     setAlternatingRowColors(true);
 
-    loadVScrollBarPosState();
+    //loadVScrollBarPosState();
```
Comment (uklotzde) [inline src/widget/wlibrarytableview.cpp]: Don't comment out code. Either delete or keep it. The next one reading this code will not know why it has been commented out and why it is still here.

File: src/widget/wtracktableview.cpp [hunk 2]
```diff
@@ -217,6 +218,11 @@ void WTrackTableView::loadTrackModel(QAbstractItemModel *model) {
         // a select() if the table is dirty.
         doSortByColumn(horizontalHeader()->sortIndicatorSection());
         return;
+    }else{
+        newModelAddr = reinterpret_cast<unsigned long>(trackModel);
```
Comment (uklotzde) [inline src/widget/wtracktableview.cpp]: This is extremely dangerous code. And it is not portable. You don't even check if a unsigned long has the same size as a pointer.

File: src/widget/wtracktableview.cpp [hunk 3]
```diff
@@ -357,6 +363,11 @@ void WTrackTableView::loadTrackModel(QAbstractItemModel *model) {
     // target though, so my hax above may not be completely unjustified.
 
     setVisible(true);
+    //restoreVScrollBarPos();
```
Comment (uklotzde) [inline src/widget/wtracktableview.cpp]: Again, what does this comment try to explain? ;)

File: src/widget/wlibrarytableview.h [hunk 4]
```diff
@@ -42,6 +55,8 @@ class WLibraryTableView : public QTableView, public virtual LibraryView {
     void loadVScrollBarPosState();
     void saveVScrollBarPosState();
 
+    QMap<QString, int> m_vScrollBarPosValues;
```
Comment (uklotzde) [inline src/widget/wlibrarytableview.h]: Why do you use QString instead of TrackModel* as your key? Pointers are valid key values.

**Comment by Be-ing** [comment]
This works when I switch between track tables and when I enter a search query then clear it, but when I switch to one of the placeholders that are shown when a library feature is selected by itself (like clicking the "Crates" header without clicking a specific crate) the scroll bar position is lost.

**Comment by Be-ing** [comment]
Thanks for taking care of this long standing issue. :)

File: src/widget/wlibrarytableview.h (hunk 4 above)
Comment (dodler) [inline src/widget/wlibrarytableview.h]: I'm quite new to C++, so I wasn't sure about that. Initially I wanted to in that way. I can still change it if needed.

**Comment by Be-ing** [comment]
@dodler do you think you'll be able to address the issues in the above comments soon? I'd like to have this fix in 2.1. Even if the position still gets lost when viewing one of the feature placeholder panels, this is still a good improvement.

**Comment by dodler** [comment]
@Be-ing yes, I will try to handle it next week.

**Comment by Be-ing** [comment]
@uklotzde, ready for merge?

**Comment by daschuer** [comment]
There are some change requests pending.

**Comment by Be-ing** [comment]
Hi @dodler, could you finish this PR?

**Comment by dodler** [comment]
Close it? I thought there are some things that should be validated. If not I'm ready.

File: src/widget/wlibrarytableview.cpp [hunk 5]
```diff
@@ -112,6 +111,21 @@ void WLibraryTableView::moveSelection(int delta) {
...
 
+void WLibraryTableView::saveVScrollBarPos(const QString key){
+    m_vScrollBarPosValues[key] = verticalScrollBar()->value();
+}
+
+void WLibraryTableView::restoreVScrollBarPos(const QString key){
+    updateGeometries();
+
+    if (m_vScrollBarPosValues.contains(key)){
+        verticalScrollBar()->setValue(m_vScrollBarPosValues[key]);
```
Comment (daschuer) [inline src/widget/wlibrarytableview.cpp]: this is a double map lookup. use find() and check for end() like here: 
https://github.com/mixxxdj/mixxx/blob/a944d1f224788e8efb41ad41bd487210766ddb11/src/control/control.cpp#L88

**Review by daschuer — CHANGES_REQUESTED** [review]
@dodler here some comments. There are also some pending commenst of @uklotzde. 
Are you able to do the requested changes? If not, ask for help. 
Thank you very much.

**Comment by daschuer** [comment]
Or did you just forget to "git push"

**Comment by dodler** [comment]
Okay, I'll finish it asap

**Comment by foss-** [comment]
@dodler Just thought this would be a useful improvement and @Be-ing pointed me to this PR. Any chance of finishing the requested changes? daschur asked to ask if help is needed.

Just a friendly ping for this PR and if you are still on this.

**Comment by dodler** [comment]
@foss- 
Hi, 
I've done some changes.

**Comment by Be-ing** [comment]
It looks like review comments have been addressed other than commenting out some debugging output. @uklotzde and/or @daschuer could you take another look at this?

**Comment by daschuer** [comment]
@dodler: I would like to merge this soon, any chance to comment the debug lines as requested? 

We need also your permission, please sign 
https://docs.google.com/a/mixxx.org/spreadsheet/viewform?formkey=dEpYN2NkVEFnWWQzbkFfM0ZYYUZ5X2c6MQ
and comment here when done. 

Thank you very much

**Comment by dodler** [comment]
@daschuer 
done

**Review by daschuer — APPROVED** [review]

**Comment by daschuer** [comment]
Thank you very much. It works nicely now. LGTM

**Comment by daschuer** [comment]
I like to add you to the contributor list in the about box inside Mixxx.
Which name and which character encoding do you like? Normally we use our real names there.

**Comment by dodler** [comment]
Artyom Lyan
any encoding

**Pull request merged**